### PR TITLE
Use kube-state-metrics jsonnet dependency #369

### PR DESCRIPTION
**Pull request by olegmayko**
Hi, This is draft PR for #369. There are few mismatches on labels, so the question is if there is a need to keep them or just follow upstream.

### CODE REVIEW

**Comment by olegmayko** [comment]
e2e tests failed, gonna check this.

File: manifests/kube-state-metrics-deployment.yaml [hunk 1]
```diff
@@ -2,71 +2,43 @@ apiVersion: apps/v1
...
-          requests:
-            cpu: 10m
-            memory: 20Mi
-      - args:
-        - --host=127.0.0.1
-        - --port=8081
-        - --telemetry-host=127.0.0.1
-        - --telemetry-port=8082
-        image: quay.io/coreos/kube-state-metrics:v1.9.3
+      - image: quay.io/coreos/kube-state-metrics
```
Comment (LiliC) [inline manifests/kube-state-metrics-deployment.yaml]: This is missing the version here, something went wrong during configuration.

File: manifests/kube-state-metrics-clusterRoleBinding.yaml [hunk 2]
```diff
@@ -9,4 +12,4 @@ roleRef:
 subjects:
 - kind: ServiceAccount
   name: kube-state-metrics
-  namespace: monitoring
+  namespace: kube-system
```
Comment (LiliC) [inline manifests/kube-state-metrics-clusterRoleBinding.yaml]: I think we want to keep the same namespace as we used so far?

**Comment by olegmayko** [comment]
yes, at the moment it deploys and configured RBAC for Kube-prometheus. The thing which is missing is ServiceMonitor. 
I've changed it and it discovered, try to figured out why 
image
it got added to job="monitoring".

**Comment by olegmayko** [comment]
hm, I pushed lib update and manifests in one commit , it was not good idea. How can I fix that this first check is not failing?

**Comment by LiliC** [comment]
Just soft reset the last commit and just do an `git add -p jsonnetfile.lock.json ` and only commit the lines about kube-state-metrics. Then `make clean` and `make generate-in-docker` again. Should fix this.

**Comment by olegmayko** [comment]
Thanks, it looks good at least from test perspective.

File: jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet [hunk 3]
```diff
@@ -1,313 +1,53 @@
-local k = import 'ksonnet/ksonnet.beta.4/k.libsonnet';
+local ksm = (import 'kube-state-metrics/kube-state-metrics.libsonnet') +
+            {
+              name:: 'kube-state-metrics',
+              namespace:: 'kube-system',
+              version:: '1.9.3',  //$._config.versions.kubeStateMetrics,
```
Comment (LiliC) [inline jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet]: I think this is 1.9.4, as the release was cut in between :)

File: manifests/kube-state-metrics-clusterRoleBinding.yaml (hunk 2 above)
Comment (LiliC) [reply manifests/kube-state-metrics-clusterRoleBinding.yaml]: We do want this to stay the same, so monitoring namespace.

File: jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet [hunk 4]
```diff
@@ -1,313 +1,53 @@
-local k = import 'ksonnet/ksonnet.beta.4/k.libsonnet';
+local ksm = (import 'kube-state-metrics/kube-state-metrics.libsonnet') +
+            {
+              name:: 'kube-state-metrics',
+              namespace:: 'kube-system',
```
Comment (LiliC) [inline jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet]: monitoring namespace here.

Comment (olegmayko) [reply jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet]: yes, updated.

File: jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet [hunk 5]
```diff
@@ -1,313 +1,53 @@
-local k = import 'ksonnet/ksonnet.beta.4/k.libsonnet';
+local ksm = (import 'kube-state-metrics/kube-state-metrics.libsonnet') +
+            {
+              name:: 'kube-state-metrics',
+              namespace:: 'monitoring',
+              version:: '1.9.4',  //$._config.versions.kubeStateMetrics,
+              image:: 'quay.io/coreos/kube-state-metrics',  //$._config.imageRepos.kubeStateMetrics + ':v' + $._config.versions.kubeStateMetrics,
```
Comment (brancz) [inline jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet]: Image needs to include the version, you can declare a local bar before name like this:

local cfg = self,

Then you can have image be what you have now “+ cfg.version”.

File: jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet [hunk 6]
```diff
@@ -1,313 +1,53 @@
...
-
-      local ksmServicePortMain = servicePort.newNamed('https-main', 8443, 'https-main');
-      local ksmServicePortSelf = servicePort.newNamed('https-self', 9443, 'https-self');
-
-      service.new('kube-state-metrics', $.kubeStateMetrics.deployment.spec.selector.matchLabels, [ksmServicePortMain, ksmServicePortSelf]) +
-      service.mixin.metadata.withNamespace($._config.namespace) +
-      service.mixin.metadata.withLabels({ 'k8s-app': 'kube-state-metrics' }) +
-      service.mixin.spec.withClusterIp('None'),
-
+    clusterRole: ksm.clusterRole,
```
Comment (brancz) [inline jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet]: No need to repeat there you can just specify the servicemonitor where the hidden config fields are specified.

File: jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet (hunk 5 above)
Comment (olegmayko) [reply jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet]: fixed.

File: jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet (hunk 6 above)
Comment (olegmayko) [reply jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet]: not sure, I understood it right. please check the last commit.

File: jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet [hunk 7]
```diff
@@ -3,7 +3,44 @@ local ksm = (import 'kube-state-metrics/kube-state-metrics.libsonnet') +
...
+                      ],
+                    },
+                    {
+                      port: 'telemetry',
+                      interval: '30s',
+                    },
+                  ],
+                },
+              },
             };
```
Comment (brancz) [inline jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet]: We can just not make this a local variable and just export this object directly.

Comment (olegmayko) [reply jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet]: fixed.

**Comment by olegmayko** [comment]
Added upstream alerts.

File: jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet (hunk 6 above)
Comment (brancz) [reply jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet]: more like, the whole file can be like this:

```jsonnet
{
  kubeStateMetrics:: (import 'kube-state-metrics/kube-state-metrics.libsonnet') + {
    local ksm = self,
    name:: 'kube-state-metrics',
    namespace:: 'monitoring',
    version:: '1.9.4',
    image:: 'quay.io/coreos/kube-state-metrics:v' + ksm.version,

    serviceMonitor: <service monitor definition using ksm variables>,
  }
}
```

Comment (brancz) [reply jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet]: fwiw this was probably a mistake: https://github.com/kubernetes/kube-state-metrics/blob/2148cb9bd5ba51d3911a93c4d61ddf084999f8c7/jsonnet/kube-state-metrics/kube-state-metrics.libsonnet#L12

We should remove the "v" prefixing and just use exactly the version string and expect people to set it with the v prefix if they want it. If you want I'd be happy about that upstream contribution :)

Comment (olegmayko) [reply jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet]: > more like, the whole file can be like this:

thanks, I will rewrite. 

> fwiw this was probably a mistake: https://github.com/kubernetes/kube-state-metrics/blob/2148cb9bd5ba51d3911a93c4d61ddf084999f8c7/jsonnet/kube-state-metrics/kube-state-metrics.libsonnet#L12

sure.

File: jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet [hunk 8]
```diff
@@ -1,316 +1,47 @@
...
+                        version:: '1.9.4',  //$._config.versions.kubeStateMetrics,
+                        image:: 'quay.io/coreos/kube-state-metrics:v' + ksm.version,
+                        serviceMonitor: {
+                          apiVersion: 'monitoring.coreos.com/v1',
+                          kind: 'ServiceMonitor',
+                          metadata: {
+                            name: ksm.name,
+                            namespace: ksm.namespace,
+                            labels: {
+                              'app.kubernetes.io/name': ksm.name,
```
[the remark this answers is not in the document]
Comment (olegmayko) [reply jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet]: right. fixed

File: jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet [hunk 9]
```diff
@@ -1,316 +1,47 @@
...
+                            },
+                          },
+                          spec: {
+                            jobLabel: 'k8s-app',
+                            selector: {
+                              matchLabels: {
+                                'app.kubernetes.io/name': ksm.name,
```
[the remark this answers is not in the document]
Comment (olegmayko) [reply jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet]: it uses ksm.commonLabels.

**Comment by brancz** [comment]
Could you squash the changes? Then lgtm! Thanks a lot for taking care of this!

**Comment by pgier** [comment]
LGTM

**Comment by LiliC** [comment]
Thanks so much @olegmayko ! 🎉

**Comment by olegmayko** [comment]
@LiliC I am happy to contribute back because I am heavily using/relying on Kube-prometheus.

**Comment by zgfh** [comment]
@olegmayko can I ask why delete resource limit  ?
image

**Comment by olegmayko** [comment]
@zgfh, that could be something which gone missing during the replacing. The resource limits are not implemented in upstream library https://github.com/kubernetes/kube-state-metrics, so if you think it's necessary you may open the ticket to add the limits there.

**Comment by lilic** [comment]
You can also just add them yourself using jsonnet in your repo :)

**Pull request merged**